tests/python/pants_test/engine/test_isolated_process.py

```diff
@@ -29,6 +29,76 @@ class ShellCat(Binary):
   def bin_path(self):
     return '/bin/cat'
 
+  def gen_argv(self, snapshots):
+    cat_file_paths = []
+    for s in snapshots:
+      cat_file_paths.extend(f.path for f in s.files)
+    # TODO(cosmicexplorer): ensure everything is a real path?
+    return (self.bin_path,) + tuple(cat_file_paths)
+
+
+class CatSourceFiles(datatype('CatSourceFiles', ['globs'])):
+
+  def __new__(cls, globs):
+
+    if not isinstance(globs, PathGlobs):
+      raise ValueError('globs should be an instance of PathGlobs')
+
+    return super(CatSourceFiles, cls).__new__(cls, globs)
+
+
+@rule(PathGlobs, [Select(CatSourceFiles)])
+def cat_source_to_globs(cat_src):
+  yield cat_src.globs
+
+
+class CatExecutionRequest(datatype('CatExecutionRequest', [
+    'shell_cat_binary',
+    'cat_source_files',
+])):
+
+  def __new__(cls, shell_cat_binary, cat_source_files):
+    if not isinstance(shell_cat_binary, ShellCat):
+      raise ValueError('shell_cat_binary should be an instance of ShellCat')
+    if not isinstance(cat_source_files, CatSourceFiles):
+      raise ValueError(
+        'cat_source_files should be an instance of CatSourceFiles')
+
+    return super(CatExecutionRequest, cls).__new__(
+      cls, shell_cat_binary, cat_source_files)
+
+
+@rule(ExecuteProcessRequest, [Select(CatExecutionRequest)])
+def cat_files_snapshotted_process_request(cat_exe_req):
+  cat_bin = cat_exe_req.shell_cat_binary
+  cat_src = cat_exe_req.cat_source_files
+  cat_files_snapshot = yield Get(Snapshot, CatSourceFiles, cat_src)
+  yield ExecuteProcessRequest.create_from_snapshot(
+    argv=cat_bin.gen_argv([cat_files_snapshot]),
+    env=tuple(),
+    snapshot=cat_files_snapshot,
+  )
+
+
+@rule(Concatted, [Select(CatExecutionRequest)])
+def cat_files_process_result_concatted(cat_exe_req):
+  # FIXME(cosmicexplorer): we should only have to run Get once here. this:
+  # yield Get(ExecuteProcessResult, CatExecutionRequest, cat_exe_req)
+  # fails because ExecuteProcessRequest is a RootRule (which shouldn't be true),
+  # but there's some work required in isolated_process.py to fix this.
+  cat_proc_req = yield Get(ExecuteProcessRequest, CatExecutionRequest, cat_exe_req)
+  cat_process_result = yield Get(ExecuteProcessResult, ExecuteProcessRequest, cat_proc_req)
+  yield Concatted(value=cat_process_result.stdout)
+
+
+def create_cat_stdout_rules():
+  return [
+    cat_source_to_globs,
+    cat_files_snapshotted_process_request,
+    cat_files_process_result_concatted,
+    RootRule(CatExecutionRequest),
```

Comment:
It's not obvious to me from the existing documentation what a RootRule is, or when one should be declared. As you seem to have a grasp of this, can you expand its documentation?

Reply:
Great idea, because I found it confusing and now I understand why I was confused. I'll make that in a separate PR.